visionmetrics/detection/confusion_matrix.py

```diff
@@ -0,0 +1,140 @@
+
+        Example:
+        predictions = [[[0, 0.9, 10, 20, 50, 100], [1, 0.8, 30, 40, 80, 120]], [[1, 0.7, 20, 30, 60, 90]]]
+        targets = [[[0, 10, 20, 50, 100], [1, 30, 40, 80, 120]], [[1, 20, 30, 60, 90]]]
+
+        Returns
+        -------
+        confusion_matrix : dict
+            Dictionary containing the counts of TP, FP, and FN, as well as details for FP reasons
+        """
+
+    def __init__(self, iou_threshold=0.5, **kwargs):
+        super().__init__(**kwargs)
+        self.iou_threshold = iou_threshold
+        self.add_state("tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_wrong_class", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_low_iou", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_no_corresponding_gt_box", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("confusion_matrix", default=[], dist_reduce_fx=None)
+
+    def update(self, predictions, targets):
+        if len(predictions) != len(targets):
+            raise ValueError("Number of predictions and targets should be the same.")
+        if not isinstance(predictions[0][0], list):
+            raise ValueError(f"Expected predictions to be a list of lists, got a list of {type(predictions[0][0])}")
+
+        self.confusion_matrix.append(self._update_confusion_matrix(predictions, targets, self.iou_threshold))
+
+    def compute(self):
+        return self.confusion_matrix[-1]
+
+    def _update_confusion_matrix(self, predictions, targets, iou_threshold):
+        for preds, gts in zip(predictions, targets):
+            gt_used = [False] * len(gts)
+
+            for pred in preds:
+                pred_class_id, score, pred_xmin, pred_ymin, pred_xmax, pred_ymax = pred
+                pred_box = [pred_xmin, pred_ymin, pred_xmax, pred_ymax]
+
+                best_iou = 0
+                best_gt_index = -1
+
+                for gt_index, gt in enumerate(gts):
+                    if gt_used[gt_index]:
+                        continue
+
+                    gt_class_id, gt_xmin, gt_ymin, gt_xmax, gt_ymax = gt
+                    gt_box = [gt_xmin, gt_ymin, gt_xmax, gt_ymax]
+
+                    if pred_class_id == gt_class_id:
+                        current_iou = self.iou(pred_box, gt_box)
+                        if current_iou > best_iou:
+                            best_iou = current_iou
+                            best_gt_index = gt_index
+
+                if best_gt_index != -1:
+                    if best_iou >= iou_threshold:
+                        self.tp += 1
+                        gt_used[best_gt_index] = True
+                    else:
+
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
```

Comment:
I think this still includes errors that are more like wrong class to me. For example, ground truth box [0,0,10,10] label 0, [90,90,100,100] label 1, prediction [90,90,100,100] label 0.  In current prediction this belongs to "fp_due_to_wrong_class". How about for boxes that don't overlap w/ any gt, they belong to "no corresponding gt box", for boxes that have overlap but labels are wrong, they belong to "wrong class"?

Reply:
I had added a note in the definition that when both "fp_due_to_wrong_class" and "fp_due_to_low_iou" conditions are satisfied, "fp_due_to_wrong_class" will be given precedence. . For example, if two boxes don't overlap and have wrong classes then the code assigns it as "fp_due_to_wrong_class" but if in the case of two non-overlapping boxes have same class they get assigned to "fp_due_to_low_iou". 

We could also flip the precedence order. I don't have a strong take on it. In that case, any two boxes with low iou (regardless of class ids) get assigned to "fp_due_to_low_iou". And FPs get assigned to "fp_due_to_wrong_class" only when iou>=thres and classes don't match. I think this is what you are referring to?

Reply:
https://github.com/microsoft/visionmetrics/pull/57